chapters/annotations.tex

```diff
@@ -336,11 +336,17 @@ \section{Annotations for Code Generation}\label{annotations-for-code-generation}
 "Evaluate" "=" ( false | true )
 \end{lstlisting}\end{synopsis}
 \begin{semantics}
-The annotation \lstinline!Evaluate! can occur in the component declaration, its type declaration, or a base-class of the type-declaration.  In the case of multiple conflicting annotations it is handled similarly to modifiers (e.g., an \lstinline!Evaluate! annotation on the component declaration takes precedence).  The annotation \lstinline!Evaluate! only has effect for a component declared with the prefix \lstinline!parameter!.
+The annotation \lstinline!Evaluate! can occur in the component declaration, its type declaration, or a base-class of the type-declaration.
+In the case of multiple conflicting annotations it is handled similarly to modifiers (e.g., an \lstinline!Evaluate! annotation on the component declaration takes precedence).
+The annotation \lstinline!Evaluate! is only allowed for parameters and constants.
 
-If \lstinline!Evaluate = true!, the model developer proposes to utilize the value for the symbolic processing. In that case, it is not possible to change the parameter value after symbolic pre-processing.
+Setting \lstinline!Evaluate = true! is only allowed for structural parameters and constants.
+For a structural parameter, it means that its value must be determined during translation, similar to a constant.
+For both parameters and constants, the model developer further proposes to utilize the value for symbolic processing.
+A constant can never be changed after translation, and it is normal for its value to be used for symbolic processing even without \lstinline!Evaluate = true!.
```

Comment:
Even if a constant cannot _be changed_ after translation it doesn't follow that the value is _known_ for symbolic processing.

Reply:
I don't understand, do you have an example?

Reply:
What are the reasons for this that you have in mind?  External function calls?

Reply:
Yes, directly or indirectly. 
Additionally we are here treating evaluating and structural as the same, which is a bit of an over-simplification. A rotational axis n={1,0,0} has a structural implication for the system of equations when we evaluate it, whereas pi=3.14159... can just be evaluated.

Reply:
That's why I think we should not add "like a constant" every time we explain an evaluated parameter.

Reply:
Sure, it should be sufficient to have "like a constant" only in a few carefully selected places.  The intention, of course, is that the refined terminology will do the rest.

Regarding over-simplification, note that the PR also introduces the term _evaluated parameter_ for the tool-dependent property of actually becoming evaluated during translation.  Regarding `pi = 3.14159…`, it would be easy to create a model where the value has a structural impact (for example, round it off and use as an array dimension).

Reply:
I don't quite see what setting Evaluate = false on a constant achieves. Is it so that it shows up in equations instead of being substituted away? What if it is needed to evaluate dimensions? Does it mean we also need to define structural constants?

Reply:
Yes, at least that's the idea.  Imagine having a big constant array; being able to prevent it from being evaluated (without having to make it a parameter) can be beneficial for performance, both during translation and at runtime.

Unfortunately, there are two important details complicating matters:
1. The sad part about constants in Modelica is that there is no guarantee that they can be evaluated during translation, and this means that this anomaly also carries over to structural parameters.  If I remember correctly, the primary challenge is external functions that can be really cumbersome to evaluate before being linked in the simulation executable.  I'm not sure that there are any obstacles that can't be overcome; it seems to me that this is just an extreme case of _quality of implementation_.
2. `Evaluate = false` is only a proposal; there is actually no way of saying that a tool must not evaluate during translation.

Hence, quite surprisingly, having a structural parameter depending on a constant with `Evaluate = false` is actually possible, and might force the `Evaluate = false` on the constant to be (legally) ignored in case the structural parameter needs to be evaluated.  If `Evaluate = false` had been imperative, it would have been natural to not allow a structural parameter to depend on such a constant.  Also note that the surprising aspect here is not introduced with structural parameters, as shown by this legal example with constants:
```
constant Real a = 5 / 2 annotation(Evaluate = false);
constant Boolean b = a > 1 annotation(Evaluate = true);
Real x = y * y + 1;
Real y = if b then 1 else x + 1;
```
(Evaluating `b` seems like a nice idea, since that allows the equations to be solved one by one instead of as a system of nonlinear equations.  However, the contradicting Evaluate annotations are allowed, and there is no rule for which one to prioritize.)

Not only does the anomaly that constants "can't" be evaluated carry over to structural parameter via expression dependency, it is just as much a problem for the structural expressions that call external functions themselves.  I am hoping that we will be able to formalize the concept of structural parameters without needing to make them better suited for translation time evaluation than constants – that we can leave this as a quality of implementation.

Reply:
I am also open to start with a separate PR about making sure all constants can be evaluated.  This would be a backwards incompatible change, but one that I think would serve us well in the end.  It would also be in line with the current state of the _Flat Modelica_ MCP: https://github.com/modelica/ModelicaSpecification/blob/MCP/0031/RationaleMCP/0031/differences.md#constant-expressions

Then, if constants are required to be evaluable, it come natural to also demand the same of the structural parameters.